Apps/Admin/Server/Services/AgentAccessService.cs

```diff
@@ -0,0 +1,212 @@
+//  limitations under the License.
+// -------------------------------------------------------------------------
+namespace HealthGateway.Admin.Server.Services
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Linq;
+    using System.Net;
+    using System.Threading.Tasks;
+    using HealthGateway.Admin.Client.Utils;
+    using HealthGateway.Admin.Common.Constants;
+    using HealthGateway.Admin.Common.Models;
+    using HealthGateway.Admin.Server.Utils;
+    using HealthGateway.Common.AccessManagement.Administration.Models;
+    using HealthGateway.Common.AccessManagement.Authentication;
+    using HealthGateway.Common.AccessManagement.Authentication.Models;
+    using HealthGateway.Common.Api;
+    using HealthGateway.Common.Constants;
+    using HealthGateway.Common.Data.ErrorHandling;
+    using HealthGateway.Common.Data.Utils;
+    using Microsoft.Extensions.Configuration;
+    using Microsoft.Extensions.Logging;
+    using Refit;
+
+    /// <summary>
+    /// Service to manage agent access to the admin website.
+    /// </summary>
+    public class AgentAccessService : IAgentAccessService
+    {
+        private const string AuthConfigSectionName = "KeycloakAdmin";
+        private readonly IAuthenticationDelegate authDelegate;
+        private readonly IKeycloakAdminApi keycloakAdminApi;
+        private readonly ILogger logger;
+        private readonly ClientCredentialsTokenRequest tokenRequest;
+        private readonly Uri tokenUri;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AgentAccessService"/> class.
+        /// </summary>
+        /// <param name="authDelegate">The OAuth2 authentication service.</param>
+        /// <param name="keycloakAdminApi">The keycloak api to access identity access.</param>
+        /// <param name="configuration">The configuration to use.</param>
+        /// <param name="logger">Injected Logger Provider.</param>
+        public AgentAccessService(
+            IAuthenticationDelegate authDelegate,
+            IKeycloakAdminApi keycloakAdminApi,
+            ILogger<AgentAccessService> logger,
+            IConfiguration configuration)
+        {
+            this.authDelegate = authDelegate;
+            this.keycloakAdminApi = keycloakAdminApi;
+            this.logger = logger;
+            _ = configuration;
+            (this.tokenUri, this.tokenRequest) = this.authDelegate.GetClientCredentialsAuth(AuthConfigSectionName);
+        }
+
+        /// <inheritdoc/>
+        [SuppressMessage("Globalization", "CA1308:Normalize strings to uppercase", Justification = "KeyCloak requires lower case username")]
+        public async Task<AdminAgent> ProvisionAgentAccessAsync(AdminAgent agent)
+        {
+            JwtModel jwtModel = this.authDelegate.AuthenticateAsSystem(this.tokenUri, this.tokenRequest);
+
+            agent.Username = agent.Username.ToLowerInvariant();
```

Comment:
1. Is this necessary? Doesn't keycloak automatically lowercase user?  If you go to kecyloak console and create a user with upper case or mixed case, it automatically lower cases it.  2.) If you are going to lowercase it here in the service, is it also being done on the front end for consistency?

Reply:
1. It's not necessary for Keycloak.
2.  I have done this because Keycloak does not give back anything but a success when the user is created, so I am returning the AdminAgent object that was passed in. Which means I need to set the username to lower case to match what it actually looks like in Keycloak.

Reply:
Removed the ToLowerInvariant and return a new AdminAgent instead since it needs the user id from Keycloak to be set.